### DISCUSSION TITLE
TFT_eSPI and ESP-NOW multi-tasking compatibility

### DISCUSSION THREAD

**Discussion by LiveRock** [Bodmer/TFT_eSPI#1488] [Q&A - General]
I am using an ESP32 with ILI9341 320x240 Touch Screen, 2 sensors and ESP-NOW, and also communication via Serial.

ESP-NOW is on core 0 and the rest on core 1.
The firmware runs ok for a few mins and then, the LCD turned inverted colors;  i.e. black becomes white, green becomes purple, etc.

I had spent the last 10 days trying all sorts of ways from Tasks Scheduler, Simple Multi-Tasking, etc. but still can't fund the source of the issue. This issue is driving me nuts...

HELP!

**Comment by LiveRock** [comment]
I am not using Task Scheduler, or vTaskPintoCore. I am just Simple Multi-Tasking
And I have also disabled the WatchDog for both cores.

**Comment by Bodmer** [comment]
I am not familiar with ESP NOW although I do know what it does.

It will not be easy to debug without a detailed knowledge of how you are using the two tasks and how they might be interacting.

Typically these messaging systems use a call back and the data will need to be stored and passed to the other processor. Only one processor can handle the TFT so the tasks need to be well partitioned.

Bear in mind that TFT_eSPI is intended for a single tasking envionment, the ESP32 is actually running an RTOS in the background so maybe these two tasks just won't cooperate to properly manage the different acitivites.

So it will need to be debugged by test and inspection to identify the root cause.

Out of interest I found this link so might have a play with ESP NOW at some point:
https://randomnerdtutorials.com/esp-now-esp32-arduino-ide/

I will move this to a discussion as there is no identifyable issue I can see with TFT_eSPI.

**Comment by LiveRock** [comment]
Thanks for responding @Bodmer 
I am searching the net for the reason why the screen went inverted colors or sometimes, total white screen.
The things is that, even when LCD go berserk, the underlying ESP-NOW is still sending data.
I just want to know why the screen go inverted. Is it a frequency issue, update issue, memory corrupted.
I checked and I have enough heap and RAM.
So, this is driving me nuts...

**Reply by Bodmer** [reply]
It screen appears to be getting bad commands for some reason and that corrupts the initialised control registers. Try reducing the SPI clock speed, keeping wires short to the display and make sure the power supply is adequate.

**Reply by LiveRock** [reply]
I have a question.
Does font size make a difference in speed (time taken to draw) when drawing on the screen.
Say, between a large font size and small font size, to draw the same text; will both draw at the same speed? Or, different?
Thanks

**Reply by LiveRock** [reply]
I reduced the (ILI9341_DRIVER)  SPI speed to 27 Mhz. Still getting white screen. Wires are short as I have a custom PCB that fit teh LCD to the ESP32. Power supply is from the laptop's USB port.

**Reply by daxliniere** [reply]
@LiveRock Peter, try a decent quality external power supply. You may be surprised at the result. I've been caught by that on a couple projects now.

**Reply by LiveRock** [reply]
Should not be the power supply as I also tested with batteries.

**Comment by Bodmer** [comment]
Characters are drawn pixel by pixel so the larger the character, the longer it takes to draw. The sketch TFT_Char_times illustrates this.

You seem to be investigating everything I would.

**Reply by LiveRock** [reply]
Thanks. Still trying and testing..
